iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -17,45 +17,153 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const capro::IdString_t& service, const capro::IdString_t& instance)
+cxx::expected<ServiceRegistry::Error> ServiceRegistry::add(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element.first == serviceDescription)
+        {
+            // Due to n:m communication we don't store twice but increase the reference counter
+            element.second++;
+            return cxx::success<>();
+        }
+    }
+    uint64_t referenceCounter = 1;
+    if (!m_serviceDescriptionVector.push_back({serviceDescription, referenceCounter}))
+    {
+        return cxx::error<Error>(Error::SERVICE_REGISTRY_FULL);
+    }
+
+    auto serviceIndex = m_serviceDescriptionVector.size() - 1;
+    m_serviceMap.insert({serviceDescription.getServiceIDString(), serviceIndex});
+    m_instanceMap.insert({serviceDescription.getInstanceIDString(), serviceIndex});
+    return cxx::success<>();
 }
 
-void ServiceRegistry::remove(const capro::IdString_t& service, const capro::IdString_t& instance)
+bool ServiceRegistry::remove(const capro::ServiceDescription& serviceDescription) noexcept
 {
-    cxx::set::remove(m_serviceMap[service].instanceSet, instance);
+    bool removedElement{false};
+    uint64_t index = 0U;
+    for (auto iterator = m_serviceDescriptionVector.begin(); iterator != m_serviceDescriptionVector.end();)
+    {
+        auto& element = m_serviceDescriptionVector[index].first;
+        if (element == serviceDescription)
+        {
+            auto& refCounter = m_serviceDescriptionVector[index].second;
+            refCounter--;
+            if (refCounter == 0)
+            {
+                m_serviceDescriptionVector.erase(iterator);
+                removedElement = true;
+            }
+            else
+            {
+                return true;
+            }
+            // There can't be more than one element
+            break;
+        }
+        index++;
+        iterator++;
+    }
+
+    auto removeIndexFromMap = [](std::multimap<capro::IdString_t, uint64_t>& map, uint64_t index) {
+        for (auto it = map.begin(); it != map.end();)
+        {
+            if (it->second == index)
+            {
+                it = map.erase(it);
+                continue;
+            }
+            else if (it->second > index)
+            {
+                // update index due to removed element
+                it->second--;
+            }
+            it++;
+        }
+    };
+
+    removeIndexFromMap(m_serviceMap, index);
```

Comment:
The lambdas should only be called when there was actually an element removed.
Could you please add `if ( removedElement ) { ...`

Otherwise we are iterating twice through some maps without doing anything.